pkg/bspatch/bspatch.go

```diff
@@ -59,34 +67,51 @@ func Reader(oldbin io.Reader, newbin io.Writer, patchf io.Reader) error {
 
 // File applies a BSDIFF4 patch (using oldfile and patchfile) to create the newfile
 func File(oldfile, newfile, patchfile string) error {
-	oldbs, err := ioutil.ReadFile(oldfile)
+	oldf, err := os.Open(oldfile)
 	if err != nil {
-		return fmt.Errorf("could not read oldfile '%v': %v", oldfile, err.Error())
+		return fmt.Errorf("could not open oldfile '%s': %v", oldfile, err)
 	}
-	patchbs, err := ioutil.ReadFile(patchfile)
+	defer oldf.Close()
+
+	newf, err := os.Create(newfile)
 	if err != nil {
-		return fmt.Errorf("could not read patchfile '%v': %v", patchfile, err.Error())
+		return fmt.Errorf("could not open or create newfile '%s': %v", newfile, err)
 	}
-	newbytes, err := patchb(oldbs, patchbs)
+
+	patchbs, err := ioutil.ReadFile(patchfile)
 	if err != nil {
-		return fmt.Errorf("bspatch: %v", err.Error())
+		return fmt.Errorf("could not read patchfile '%s': %v", patchfile, err)
 	}
-	if err := ioutil.WriteFile(newfile, newbytes, 0644); err != nil {
-		return fmt.Errorf("could not create newfile '%v': %v", newfile, err.Error())
+
+	newfw := bufio.NewWriterSize(newf, writeBufferSize)
+	err = patchStream(oldf, newfw, patchbs)
+
+	if err != nil {
+		newf.Close()
+		os.Remove(newfile)
+		return fmt.Errorf("bspatch: %v", err)
 	}
+
+	newfw.Flush()
+	newf.Close()
 	return nil
 }
 
-func patchb(oldfile, patch []byte) ([]byte, error) {
-	oldsize := len(oldfile)
-	var newsize int
-	header := make([]byte, 32)
-	buf := make([]byte, 8)
-	var lenread int
-	var i int
-	ctrl := make([]int, 3)
+func patchStream(oldf io.ReadSeeker, newf io.Writer, patch []byte) error {
+	const HeaderLen = 32
 
-	f := bytes.NewReader(patch)
+	// Reused container vars
+	var lenread int64
+	var errmsg string
+	header := make([]byte, HeaderLen)
+	hdbuf := make([]byte, 8)
+	ctrip := make([]int64, 3)
+	// ctrl triple indices
+	var x, y, z = 0, 1, 2
```

Comment:
I find this more confusing than simply using 0, 1, 2. Better could be

```
type ctrl [3]int64
func (c ctrl) sum() int64
func (c ctrl) copy() int64
func (c ctrl) seek() int64
```

Reply:
ah yeah, think i've been holding the comment notation in my head too long. will clarify this